Temporal acceptance tests

fickludd opened the pull request:
This pull request contains mostly acceptance test refactoring and additions, testing both spatial and temporal indexing capabilities. Some minor bugs were detected and fixed, and others cardified. In particular, we ran into a bug with the spatial index which was deemed to hard to solve here, or possibly we are just missing something. 

In the final commit we decided to completely refactor how `NodeIndexSeekPipe` execute index seeks, because this was to us very hard to understand. All logic around this is now encapsulated in the new trait `NodeIndexSeeker`, which directly creates the kernel API `IndexQuery` that is used to perform the index seeks. This is possible because we are new versioning cypher on the logical plan.

~~Builds on temporal index IV (#11181), first commit is `split spatial index test classes into separate files`.~~

fickludd committed:
split spatial index test classes into separate files

fickludd committed:
Clean up SpatialIndexAcceptanceTest and break out support functionality

fickludd committed:
Repurpose SpatialIndexAcceptanceTest to test index persistence for all types

fickludd committed:
Fix issues in temporal and spatial exists queries

fickludd committed:
New temporal index query acceptance tests

fickludd committed:
Refactor semantic index acceptance test for more extensibility

fickludd committed:
Introduce increased generated testing of cypher using index seeks

This commit extends SemanticIndexAcceptanceTest to test Point and
Date in addition to the previous Long/Double/String. In doing this,
several minor problems were found for both code paths and corrected.

Note that this commit further extends the shaky index query code around
TransactionBoundQueryContext, which extends all the way into
PropertyValueComparison and other places. This was done to get something
which passes tests, but this area is in dire need of a rewrite. The
other temporal types should not be stacked on top of that mess.

fickludd committed:
Refactor interpreted/slotted runtime index seeks

The logic for converting logical index seeks into executable
queries has been gathered from various places into the new
trait NodeIndexSeeker. This trait converts logical index seeks,
range scans, special string seeks and distance queries into the
Kernel API IndexQuery classes, which allows direct querying.

This improves the type definitions in touched QueryContext methods,
makes the functionality easier to understand, and reduces object
creation. It also gives a big reduction of TransactionBoundQueryContext.

Note that this refactoring was enabled but the switch to cypher
versioning on the logical plan level, which allows us to utilize
kernel types in runtime code.

fickludd committed:
Fix tests which created invalid points

fickludd committed:
Remove redundant test, and migrate another to the correct method

fickludd committed:
Update more broken spatial fusion tests

sherfert commented on community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/SemanticIndexAcceptanceTest.scala:
```diff
@@ -22,38 +22,72 @@ package org.neo4j.cypher.internal.compiler.v3_4
+    for {
+      dimension <- Gen.oneOf(allCRSDimensions)
+      coordinates <- Gen.listOfN(dimension, arbitrary[Double])
+      crs <- Gen.oneOf(allCRS(dimension))
```
dimension and crs need to be compatible. That is, you cannot use dimension 2 with CRS "Cartesian-3D"

Lojjs replied:
I think that is already handled. `allCRS: Map[Int, Array[CoordinateReferenceSystem]] = CoordinateReferenceSystem.all().toArray.groupBy(_.getDimension)` above groups the crs after their nbr of dimensions and then one with the right nbr ofg dimensions is chosen with `Gen.oneOf(allCRS(dimension))`  

(Note. this is a reply to Satia's comment)

sherfert commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionIndexPopulatorTest.java:
```diff
@@ -64,7 +64,9 @@ public void setup() throws Exception
-        for ( CoordinateReferenceSystem crs : asList( CoordinateReferenceSystem.WGS84, CoordinateReferenceSystem.Cartesian ) )
+        for ( CoordinateReferenceSystem crs : asList( CoordinateReferenceSystem.WGS84,
+                                                      CoordinateReferenceSystem.Cartesian,
+                                                      CoordinateReferenceSystem.Cartesian_3D ) )
```
What is the reason WGS84-3D is excluded?

fickludd replied:
No clue. This test used to have a wrongly labelled `Cartesian_3D`, this addition is merely fixing a previously broken setup. 

This test will be deleted soon anyway by the spatial rewrite.

Lojjs commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/TemporalIndexAcceptanceTest.scala:
```diff
@@ -0,0 +1,60 @@
+    assertSeekMatchFor(value, node)
+  }
+
+  def assertRangeScan(v1: Value, v2: Value, v3: Value, v4: Value): Unit = {
```
I think you should extend this test somewhat, at least checking `>=` and `<=` also.

fickludd replied:
True that, I think we got stuck with some bug and then forgot about it. Will add more here!

Lojjs commented on community/bolt/src/test/java/org/neo4j/bolt/v1/messaging/AuthTokenValuesWriterTest.java:
```diff
@@ -80,7 +80,7 @@ public void shouldFailToWriteRelationship()
     {
         AuthTokenValuesWriter writer = new AuthTokenValuesWriter();
-        PointValue value = pointValue( CoordinateReferenceSystem.WGS84, new double[42] );
+        PointValue value = pointValue( CoordinateReferenceSystem.WGS84, new double[2] );
```
Out of curiosity, why didn't `new double[42]` work?

fickludd replied:
`WGS84` only has 2 dimensions...

Lojjs replied:
Haha didn't realise that was the nbr of dimensions

Lojjs commented on community/bolt/src/test/java/org/neo4j/bolt/v2/messaging/Neo4jPackV2Test.java:
```diff
@@ -335,8 +337,11 @@ private PointValue randomPoint3D( int index )
-        CoordinateReferenceSystem crs = index % 2 == 0 ? WGS84 : Cartesian;
-        return pointValue( crs, random().doubles( dimension, Double.MIN_VALUE, Double.MAX_VALUE ).toArray() );
+        CoordinateReferenceSystem crs =
+                index % 2 == 0 ?
```
For increased readability I would prefer the outer `if` to **not** be written with `x ? y : z` notation

Lojjs commented on community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/SemanticIndexAcceptanceTest.scala:
```diff
@@ -22,38 +22,72 @@ package org.neo4j.cypher.internal.compiler.v3_4
       ValueSetup[DoubleValue]("doubles", doubleGen, x => x.minus(0.1), x => x.plus(0.1)),
-      ValueSetup[TextValue]("strings", textGen, changeLastChar(c => (c - 1).toChar), changeLastChar(c => (c + 1).toChar))
+      ValueSetup[TextValue]("strings", textGen, changeLastChar(c => (c - 1).toChar), changeLastChar(c => (c + 1).toChar)),
+//      ValueSetup[PointValue]("points", pointGen, modifyPoint(_ - 0.1), modifyPoint(_ + 0.1)), // TODO: here is a bug
```
Have you cardified this bug?

fickludd replied:
Yupp.

Lojjs commented on community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/SemanticIndexAcceptanceTest.scala:
```diff
@@ -22,38 +22,72 @@ package org.neo4j.cypher.internal.compiler.v3_4
+
+  def dateGen: Gen[DateValue] =
+    for {
+      epochDays <- arbitrary[Int] //Gen.chooseNum(-MAX_EPOCH_DAYS+1, MAX_EPOCH_DAYS-1)?
```
I think you should expand, remove or/and cardify this comment

Lojjs commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/IndexingTestSupport.scala:
```diff
@@ -0,0 +1,81 @@
+  }
+
+  private def testIndexedRead(query: String, params: Map[String, AnyRef], wantedOperator: String, expected: Seq[Node]): Unit = {
+    val result = graph.execute(query, params)
```
We are only testing with `slotted`. Why are we not utilising `CypherComparisonSupport` and running `executeWith` here instead?

fickludd replied:
We forced it to slotted because of the bug in passing `Point` as a compiled runtime parameter. The running with only `graph.execute` we inherited from the previous test implementation, I guess it is because the test was mostly about testing index value persistence on restarts, which should not be dependent on the runtime/planner.

fickludd replied:
...but you are right. Migrated `TemporalIndexAcceptanceTest` to use `CypherComparisonSupport`.

Lojjs commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/TemporalIndexReader.java:
```diff
@@ -95,6 +87,7 @@ public void query( IndexProgressor.NodeValueClient cursor, IndexOrder indexOrder
         IndexQuery predicate = predicates[0];
         if ( predicate instanceof ExistsPredicate )
         {
+            loadAll();
```
Nice fix! :)

Lojjs commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/ExceptionTranslatingQueryContext.scala:
```diff
@@ -228,9 +228,6 @@ class ExceptionTranslatingQueryContext(val inner: QueryContext) extends QueryCon
   override def getRelationshipFor(relationshipId: Long, typeId: Int, startNodeId: Long, endNodeId: Long): RelationshipValue =
     translateException(inner.getRelationshipFor(relationshipId, typeId, startNodeId, endNodeId))
 
-  override def indexSeekByRange(index: IndexReference, value: Any) =
```
I don't understand why you are able to remove `indexSeekByRange` everywhere? Can you explain?

fickludd replied:
We now use `indexSeek` for most types of queries, and encode the range query as an `IndexQuery` object directly in `NodeIndexSeeker`.

Lojjs commented on community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/pipes/NodeIndexSeekPipe.scala:
```diff
@@ -46,15 +46,12 @@ case class NodeIndexSeekPipe(ident: String,
+    val indexReference = reference(state.query)
     val baseContext = state.createOrGetInitialContext(executionContextFactory)
-    val resultNodes = indexQuery(valueExpr, baseContext, state, index, label.name, propertyKeys.map(_.name))
+    val resultNodes: Iterator[NodeValue] = indexSeek(state, indexReference, baseContext)
```
Unnecessary type annotation

Lojjs commented on enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/NodeIndexSeekSlottedPipe.scala:
```diff
@@ -53,14 +53,12 @@ case class NodeIndexSeekSlottedPipe(ident: String,
+    val indexReference = reference(state.query)
     val baseContext = state.createOrGetInitialContext(executionContextFactory)
-    val resultNodes = indexQuery(valueExpr, baseContext, state, index, label.name, propertyKeys.map(_.name))
+    val resultNodes: Iterator[NodeValue] = indexSeek(state, indexReference, baseContext)
```
Unnecessary type annotation

Lojjs commented on community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/pipes/NodeIndexSeeker.scala:
```diff
@@ -0,0 +1,201 @@
+
+      case _: SeekByRange =>
+        val indexQueries = computeIndexQueries(state, baseContext)
+        indexQueries.flatMap(query => state.query.indexSeek(indexReference, query)).toIterator
```
This case looks identical to `ExactSeek`, so you can simplify this code somewhat by

```
case _: ExactSeek =>
case _: SeekByRange =>
    ...
```

fickludd replied:
Well I agree that they are the same, but there is no fallthrough in scala, so that won't work. I guess we could just split it up into `LockingUniqueIndexSeek` and everything else. Will fix!

Lojjs reviewed (changes requested):
Overall very good work! I have some questions and comments though. I haven't looked to deep into all details of the index seek refactoring, but at a glance it looks good.

fickludd committed:
Enrich range-scan acceptance tests and fix minor discovered issues

Also:
 - rebase fixes
 - migrate IndexingTestSupport to CypherComparisonSupport

fickludd committed:
Add acceptance testing of half-ranges

fickludd committed:
Misc readability improvements

Lojjs commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/IndexingTestSupport.scala:
```diff
@@ -51,31 +54,50 @@ trait IndexingTestSupport extends ExecutionEngineFunSuite with CypherComparisonS
+    if (cypherComparisonSupport) {
+      val result =
+        executeWith(
+          TestConfiguration(
```
I'm pretty sure this TestConfiguration can be written as Configs.Interpreted - Configs.OldAndRule

Lojjs commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/IndexingTestSupport.scala:
```diff
@@ -51,31 +54,50 @@ trait IndexingTestSupport extends ExecutionEngineFunSuite with CypherComparisonS
+          query,
+          params = params,
+          planComparisonStrategy = ComparePlansWithAssertion(_ should useOperators(wantedOperator),
+                                                             expectPlansToFail = Configs.AllRulePlanners)
```
` expectPlansToFail = Configs.AllRulePlanners` is not needed because you don't expect any RulePlanners to succeed so their plans will not be checked anyway

Lojjs replied:
`expectPlansToFail = Configs.AllRulePlanners` is not needed because you don't expect any RulePlanners to succeed so their plans will not be checked anyway

Lojjs commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/TemporalIndexAcceptanceTest.scala:
```diff
@@ -42,6 +44,30 @@ class TemporalIndexAcceptanceTest extends IndexingTestSupport {
+      LocalTimeValue.localTime(10006))
+
+    assertRangeScan(
+      TimeValue.time(10000, ZoneId.of("Europe/Stockholm")),
```
Will the `ZoneId.of` return some zone offset? Because `TimeValue` should not work with named timezones...

fickludd replied:
Weeeell, it used to convert into an offset later. We changed it to only permit offsets in the factory method as expected.

Lojjs commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/IndexingTestSupport.scala:
```diff
@@ -0,0 +1,109 @@
+    if (cypherComparisonSupport) {
+      val result =
+        executeWith(
+          TestConfiguration(
```
This TestConfiguration can be more compactly written as `Configs.Interpreted - Configs.OldAndRule`

OliviaYtterbrink committed:
Fix equals for zoned date time and added MIN/MAX

fickludd committed:
Make TimeValue.time() require a ZoneOffset

fickludd committed:
Change TimeValue.writeTo and time() to use nanosPerDayLocal

We previously used nanosPerDayUTC, but this seemed less useful
as the value typically is specified in local time. To arrive
at a valid nanosPerDayUTC we might therefore overflow the valid
range as we add the offset to the nanosPerDayLocal, which would
a) require expensive modulus computations and b) mean that the
nanosPerDayUTC would not always compare as needed for TimeValue
comparison semantics. From a valid nanosPerDayLocal we can easily
compute an potentially invalid nanosPerDayUTC using the offset,
that can be used for semantically correct comparison.

fickludd committed:
Fix bug in Time indexing with overflowing UTC time

fickludd committed:
Fix DateTimeValue equals and compareTo; move TimeZones to values module

DateTimeValue now correctly marks renamed time zones as being equal.

fickludd committed:
DateTime sorting of offsets and named zones

Always sort DateTimes with a offset before one with a named
zone if they have the same instant and local time.

fickludd committed:
Randomized testing of all temporal types which can be range queried

fickludd committed:
cleanups around TimeValue

fickludd committed:
Fix ZonedDateTime index issue with negative offsets, offset now in minutes

fickludd committed:
Add MIN/MAX values for temporal value types

fickludd committed:
cleanup configuration in IndexingTestSupport

Lojjs commented on community/values/src/main/java/org/neo4j/values/storable/DateTimeValue.java:
```diff
@@ -427,10 +435,39 @@ public boolean equals( Value other )
+        int cmp = Long.compare( value.toEpochSecond(), that.toEpochSecond() );
+        if ( cmp == 0 )
+        {
+            cmp = value.toLocalTime().getNano() - that.toLocalTime().getNano();
```
Why do we have to check the nanos here, aren't they included in `that.toLocalTime()` and part of the next comparison?

Lojjs commented on community/values/src/main/java/org/neo4j/values/storable/DateTimeValue.java:
```diff
@@ -427,10 +435,39 @@ public boolean equals( Value other )
     {
-        DateTimeValue other = (DateTimeValue) otherValue;
-        return value.compareTo( other.value );
+        return !(thisZone instanceof ZoneOffset) &&
```
I find this a bit strange. Consider the example `thisZone = +05:00` and `thatZone = 'Europe/Stockholm'`. Since thisZone is a ZoneOffset, the method will return false which mean they are considered the same timezone?? Maybe it is just the name of the method that confuses me...

Lojjs commented on community/values/src/main/java/org/neo4j/values/storable/TimeZones.java:
```diff
@@ -122,5 +124,13 @@ else if ( line.startsWith( "#" ) )
+            String oldName = entry.getKey();
+            String newName = TIME_ZONE_SHORT_TO_STRING.get( entry.getValue() );
+            Short newNameId = tempStringToShort.get( newName );
+            TIME_ZONE_STRING_TO_SHORT.put( oldName, newNameId );
```
I think this can be improved by instead of `tempStringToShort` having a `tempStringToString` that just saves the old and new name for those who actually changed name. Then the loop will only be over the renamed zones and its inner can be rewritten like

```
String oldName = entry.getKey();
String newName = entry.getValue();
Short newNameId = TIME_ZONE_SHORT_TO_STRING.get(newName);
TIME_ZONE_STRING_TO_SHORT.put( oldName, newNameId );
```

fickludd replied:
I agree, will fix this later.

Lojjs commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/ZonedTimeSchemaKey.java:
```diff
@@ -67,9 +72,9 @@ public void initValueAsHighest()
+        if ( compare == 0 )
         {
-            compare = Values.COMPARATOR.compare( asValue(), other.asValue() );
+            compare = Integer.compare( zoneOffsetSeconds, other.zoneOffsetSeconds );
```
Was the `Values.COMPARATOR.compare` thing we did initially unnecessary?

Lojjs commented on community/kernel/src/test/java/org/neo4j/kernel/api/index/SimpleIndexAccessorCompatibility.java:
```diff
@@ -311,6 +312,16 @@ public void testIndexRangeSeekByTimeWithDuplicates() throws Exception
+            testIndexRangeSeekWithDuplicates( time( 20, 31, 53, 4, ZoneOffset.of("+17:02") ),
+                                              time( 20, 31, 54, 3, ZoneOffset.of("+17:02") ),
+                                              time( 19, 31, 54, 2, UTC ),
+                                              time( 18, 23, 27, 1, ZoneOffset.of("-18:00") ) );
```
Doesn't the local time in the last case become `00:23:27` i.e. first in the range?

Lojjs commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/ZonedDateTimeLayout.java:
```diff
@@ -84,19 +84,19 @@ public void readKey( PageCursor cursor, ZonedDateTimeSchemaKey into, int keySize
     private int asZoneOffset( int encodedZone )
     {
-        return encodedZone;
+        return (short) encodedZone;
```
If you should benefit from returning a `short` here shouldn't the return type of the method also be `short`?

Lojjs commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/ZonedDateTimeLayout.java:
```diff
@@ -70,7 +70,7 @@ public void writeKey( PageCursor cursor, ZonedDateTimeSchemaKey key )
         else
         {
-            cursor.putInt( key.zoneOffsetSeconds );
+            cursor.putInt( key.zoneOffsetMinutes & ZONE_ID_MASK );
```
Why `& ZONE_ID_MASK` ?

Lojjs reviewed (commented):
I have looked through your new stuff now, great work! My new feedback is mostly questions about things I wasn't sure about/didn't understand. Let's go through it together tomorrow and hopefully merge it afterwards!

Lojjs commented on community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/SemanticIndexAcceptanceTest.scala:
```diff
@@ -19,73 +19,153 @@
-    }
+  def timeGen: Gen[TimeValue] =
+    for {
+      nanosOfDay <- Gen.chooseNum(1000000000, 86399999999999L - 1000000000)
```
Where does this range come from? Why are you not checking values closer to 0? Maybe add a comment about it.

fickludd replied:
Yes, let's improve readability here and add a comment or named constants.

Lojjs commented on community/values/src/main/java/org/neo4j/values/storable/DateTimeValue.java:
```diff
@@ -62,6 +62,9 @@
 public final class DateTimeValue extends TemporalValue<ZonedDateTime,DateTimeValue>
 {
+    public static final DateTimeValue MIN_VALUE = new DateTimeValue( ZonedDateTime.of( LocalDateTime.MIN, ZoneOffset.MIN ) );
+    public static final DateTimeValue MAX_VALUE = new DateTimeValue( ZonedDateTime.of( LocalDateTime.MAX, ZoneOffset.MAX ) );
```
Have you double checked that this doesn't overflow, because I think it is saved in UTC? Or was that only the TimeValue before you changed that in another commit of this PR?

Lojjs commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/ZonedTimeSchemaKey.java:
```diff
@@ -82,9 +90,9 @@ public String toString()
+    public void writeTime( long nanosOfDayLocal, int offsetSeconds )
     {
-        this.nanosOfDayUTC = nanosOfDayUTC;
+        this.nanosOfDayUTC = nanosOfDayLocal - offsetSeconds * 1_000_000_000L;
```
Perhaps you want to use your new `NANOS_PER_SECOND` here instead of hardcode `1_000_000_000L`?

fickludd replied:
yupp.

Lojjs commented on community/values/src/main/java/org/neo4j/values/storable/DateTimeValue.java:
```diff
@@ -429,10 +435,42 @@ public boolean equals( Value other )
+                    ZoneId thatZone = that.getZone();
+                    boolean thisIsOffset = thisZone instanceof ZoneOffset;
+                    boolean thatIsOffset = thatZone instanceof ZoneOffset;
+                    cmp = Boolean.compare( thatIsOffset, thisIsOffset ); // offsets before named zones
```
I can't find this rule about offsets before named timezones in the CIP, is it documented somewhere?

Lojjs committed:
Fix test failure

fickludd commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/IndexingTestSupport.scala:
```diff
@@ -80,15 +80,15 @@ trait IndexingTestSupport extends ExecutionEngineFunSuite with CypherComparisonS
+//            Versions(Versions.V3_4, Versions.V3_3, Versions.Default),
+//            Planners(Planners.Cost, Planners.Default),
+//            Runtimes(Runtimes.Interpreted, Runtimes.Slotted, Runtimes.Default)
+//          ),
```
Ghostlines!

fickludd committed:
Do not let index seeks throw on impossible predicates

If a predicate is statically guaranteed to be impossible, we should
still treat this as a user error and give a nice error message. This
should not be dependent on the existence or not of an index, and should
be done during linting as soon as type inference is complete.

fickludd committed:
Minor readability improvements

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/TemporalIndexCache.java:
```diff
@@ -151,7 +151,7 @@ T date() throws E
         if ( date == null )
         {
             date = factory.newDate();
```
So these factories can return `null` sometimes? Are they generally costly to call, if so perhaps a flag for knowing when they have been called could be in order

fickludd replied:
The only one which returns `null` is the `TemporalIndexReader`, which is not allowed to create new accessors (and therefore indexes) for temporal types which do not have any indexes values. In this case calling the factory is super-cheap: essentially only a switch and a null-check, so we saw no particular reason to be clever there. Will still need to check on every query though, because a new temporal value might have been added since the last query, causing the index to now exist.

We did think about refactoring here for nicer type-safety around `uncheckedSelect`, which is only used from the reader, where we statically know it will not throw, but I guess that's a slightly different concern.
